File: crates/app-core/src/commands/input.rs
use yaya_provider_api::{BinaryAsset, ProviderView};

use crate::{AppCore, AppError};

impl AppCore {
    /// Resolve raw user input into a provider view via the registry.
    pub async fn inspect_source(&self, source: String) -> Result<ProviderView, AppError> {
        Ok(self.runtime.inspect(source).await?)
    }

    /// Fetch a small asset (cover image …) through the owning provider's
    /// network context.
    pub async fn fetch_provider_asset(
        &self,
        provider_id: &str,
        url: &str,
    ) -> Result<BinaryAsset, AppError> {
        Ok(self.runtime.fetch_provider_asset(provider_id, url).await?)
    }
}

File: src/services/api/input.ts
import type { BinaryAsset, ProviderView } from '../../types'
import { transport } from '../transport'

export function inspectSource(source: string) {
  return transport<ProviderView>(
    'inspect_source',
    { source },
    '/inspect',
    { method: 'POST', body: JSON.stringify({ source }) },
  )
}

const assetCache = new Map<string, Promise<string>>()

/** Fetch a provider-context image and return it as a data URL (cached). */
export function providerImage(providerId: string, url: string): Promise<string> {
  if (!/^https?:\/\//i.test(url))
    return Promise.resolve(url)
  const key = `${providerId}\0${url}`
  const existing = assetCache.get(key)
  if (existing)
    return existing
  const request = transport<BinaryAsset>(
    'fetch_provider_asset',
    { providerId, url },
    `/providers/${encodeURIComponent(providerId)}/asset?url=${encodeURIComponent(url)}`,
  )
    .then(asset => `data:${asset.contentType};base64,${asset.bytes}`)
    .catch((error) => {
      assetCache.delete(key)
      throw error
    })
  assetCache.set(key, request)
  return request
}
